kernel/gic/src/dist.rs

```diff
@@ -0,0 +1,205 @@
+use super::MmioPageOfU32;
+use super::U32BYTES;
+use super::TargetCpu;
+use super::IntNumber;
+use super::Enabled;
+use super::read_array;
+use super::write_array;
+use super::TargetList;
+
+mod offset {
+    use super::U32BYTES;
+    pub const CTLR:      usize = 0x000 / U32BYTES;
+    pub const IGROUPR:   usize = 0x080 / U32BYTES;
+    pub const ISENABLER: usize = 0x100 / U32BYTES;
+    pub const ICENABLER: usize = 0x180 / U32BYTES;
+    pub const ITARGETSR: usize = 0x800 / U32BYTES;
+    pub const SGIR:      usize = 0xf00 / U32BYTES;
+    /// This one is on the 6th page
+    pub const P6IROUTER: usize = 0x100 / U32BYTES;
+}
+
+// enable group 0
+// const CTLR_ENGRP0: u32 = 0b01;
+
+// enable group 1
+const CTLR_ENGRP1: u32 = 0b10;
+
+// Affinity Routing Enable, Non-secure state.
+const CTLR_ARE_NS: u32 = 1 << 5;
+
+// bits [24:25]: target mode
+//   1 = all other PEs
+//   0 = use target list
+const SGIR_TARGET_ALL_OTHER_PE: u32 = 1 << 24;
+
+// const GROUP_0: u32 = 0;
+const GROUP_1: u32 = 1;
+
+// bit 15: which interrupt group to target
+const SGIR_NSATT_GRP0: u32 = 0 << 15;
+
+fn assert_cpu_bounds(target: &TargetCpu) {
+    if let TargetCpu::Specific(cpu) = target {
+        assert!(*cpu < 8, "affinity routing is disabled; cannot target a CPU with id >= 8");
+    }
+}
+
+/// Initializes the distributor by enabling forwarding
+/// of group 1 interrupts
+///
+/// Return value: whether or not affinity routing is
+/// currently enabled for both secure and non-secure
+/// states.
+pub fn init(registers: &mut MmioPageOfU32) -> Enabled {
+    let mut reg = registers[offset::CTLR];
+    reg |= CTLR_ENGRP1;
+    registers[offset::CTLR] = reg;
+
+    // Return value: whether or not affinity routing is
+    // currently enabled for both secure and non-secure
+    // states.
+    reg & CTLR_ARE_NS > 0
+}
+
+/// Will that interrupt be forwarded by the distributor?
```

Comment:
Used your doc + documented the use of "Forwarding" as so:
```rust
//! The term "Forwarding" is sometimes used in this crate.
//! This is because the Distributor, Redistributor and CPU interface are
//! chained in the controller. The distributor and the redistributor are
//! configured by the code of this crate to either allow (forward)
//! interrupts or disallow (discard) them.
```
As this is distributor code, ints are forwarded to the redistributor.

Reply:
great!